The full build log shows
```
>> Building shared library tmp/el8_amd64_gcc12/src/DataFormats/Provenance/src/DataFormatsProvenance/libDataFormatsProvenance.so
...
In member function 'deallocate',
    inlined from 'deallocate' at /cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/include/c++/12.3.1/bits/allocator.h:200:35,
    inlined from 'deallocate' at /cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/include/c++/12.3.1/bits/alloc_traits.h:496:23,
    inlined from '_M_destroy' at /cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/include/c++/12.3.1/bits/basic_string.h:300:34,
    inlined from '_M_dispose' at /cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/include/c++/12.3.1/bits/basic_string.h:294:14,
    inlined from '_M_dispose' at /cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/include/c++/12.3.1/bits/basic_string.h:291:7,
    inlined from '__dt_base ' at /cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/include/c++/12.3.1/bits/basic_string.h:803:19,
    inlined from 'compactForm_' at src/DataFormats/Provenance/src/Hash.cc:25:5:
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/include/c++/12.3.1/bits/new_allocator.h:158:33: warning: 'operator delete' called on unallocated object 'temp' [-Wfree-nonheap-object]
  158 |         _GLIBCXX_OPERATOR_DELETE(_GLIBCXX_SIZED_DEALLOC(__p, __n));
      |                                 ^
src/DataFormats/Provenance/src/Hash.cc: In function 'compactForm_':
src/DataFormats/Provenance/src/Hash.cc:22:18: note: declared here
   22 |       value_type temp(hash);
      |                  ^
```
which seems to be coming from LTO, but from code not touched by this PR. With @Dr15Jones we have no clue what could be causing the warning.

@smuzaffar Is this warning not being flagged in the PR test summary because the `Hash.cc` was not touched by this PR?

I also noticed this warning was shown a second time later in the build log for the same shared library (including the same `Building shared library` line). @smuzaffar Do you know why that happens?

--- a/DataFormats/Provenance/interface/ProductProvenance.h
+++ b/DataFormats/Provenance/interface/ProductProvenance.h
@@ -39,8 +39,7 @@ namespace edm {
     ParentageID const& parentageID() const { return parentageID_; }
     Parentage const& parentage() const;
 
-    void set(ParentageID id) { parentageID_ = std::move(id); }
-    ParentageID moveParentageID() { return std::move(parentageID_); }
+    void set(ParentageID const& id) { parentageID_ = id; }
 
   private:
     BranchID branchID_;

--- a/DataFormats/Provenance/interface/ProductProvenanceLookup.h
+++ b/DataFormats/Provenance/interface/ProductProvenanceLookup.h
@@ -33,7 +33,7 @@ namespace edm {
     ProductProvenanceLookup& operator=(ProductProvenanceLookup const&) = delete;
 
     ProductProvenance const* branchIDToProvenance(BranchID const& bid) const;
-    void insertIntoSet(ProductProvenance provenanceProduct) const;
+    void insertIntoSet(ProductProvenance const& provenanceProduct) const;
     ProductProvenance const* branchIDToProvenanceForProducedOnly(BranchID const& bid) const;
 
     void update(edm::ProductRegistry const&);
@@ -78,8 +78,8 @@ namespace edm {
 
       bool isParentageSet() const noexcept { return isParentageSet_.load(std::memory_order_acquire); }
 
-      void threadsafe_set(ParentageID id) const {
-        provenance_.set(std::move(id));
+      void threadsafe_set(ParentageID const& id) const {
+        provenance_.set(id);
         isParentageSet_.store(true, std::memory_order_release);
       }
 

--- a/DataFormats/Provenance/src/CompactHash.cc
+++ b/DataFormats/Provenance/src/CompactHash.cc
@@ -32,7 +32,7 @@ namespace edm {
 
     std::array<unsigned char, 16> fromHex_(std::string_view v) {
       cms::MD5Result temp;
-      temp.fromHexifiedString(std::string(v));
+      temp.fromHexifiedString(v);
       auto hash = temp.compactForm();
       std::array<unsigned char, 16> ret;
       std::copy(hash.begin(), hash.end(), ret.begin());

--- a/DataFormats/Provenance/src/Parentage.cc
+++ b/DataFormats/Provenance/src/Parentage.cc
@@ -1,7 +1,8 @@
 #include "DataFormats/Provenance/interface/Parentage.h"
 #include "FWCore/Utilities/interface/Digest.h"
-#include <ostream>
+#include <charconv>
 #include <sstream>
+//#include <cassert>
 
 /*----------------------------------------------------------------------
 
@@ -15,14 +16,17 @@ namespace edm {
   Parentage::Parentage(std::vector<BranchID>&& parents) : parents_(std::move(parents)) {}
 
   ParentageID Parentage::id() const {
-    std::ostringstream oss;
+    //10 is the maximum number of digits for a 2^32 number
+    std::array<char, 10 + 1> buf;
+    cms::Digest md5alg;
     for (auto const& parent : parents_) {
-      oss << parent << ' ';
+      //assert(start < end);
+      auto res = std::to_chars(buf.data(), buf.data() + buf.size(), parent.id());
+      //assert(res.ec == std::errc());
+      *res.ptr = ' ';
+      md5alg.append(buf.data(), res.ptr - buf.data() + 1);
     }
-
-    std::string stringrep = oss.str();
-    cms::Digest md5alg(stringrep);
-    ParentageID id(md5alg.digest().toString());
+    ParentageID id(md5alg.digest().bytes);
     return id;
   }
 

--- a/DataFormats/Provenance/src/ProductProvenanceLookup.cc
+++ b/DataFormats/Provenance/src/ProductProvenanceLookup.cc
@@ -39,7 +39,7 @@ namespace edm {
     setupEntryInfoSet(iReg);
   }
 
-  void ProductProvenanceLookup::insertIntoSet(ProductProvenance entryInfo) const {
+  void ProductProvenanceLookup::insertIntoSet(ProductProvenance const& entryInfo) const {
     //NOTE:do not read provenance here because we only need the full
     // provenance when someone tries to access it not when doing the insert
     // doing the delay saves 20% of time when doing an analysis job
@@ -53,7 +53,7 @@ namespace edm {
       throw edm::Exception(edm::errors::LogicError) << "ProductProvenanceLookup::insertIntoSet passed a BranchID "
                                                     << entryInfo.branchID().id() << " that has not been pre-registered";
     }
-    itFound->threadsafe_set(entryInfo.moveParentageID());
+    itFound->threadsafe_set(entryInfo.parentageID());
   }
 
   ProductProvenance const* ProductProvenanceLookup::branchIDToProvenance(BranchID const& bid) const {

--- a/DataFormats/Provenance/test/Parentage_t.cpp
+++ b/DataFormats/Provenance/test/Parentage_t.cpp
@@ -32,4 +32,35 @@ TEST_CASE("test Parentage", "[Parentage]") {
   edm::ParentageID id4 = ed4.id();
   CHECK(ed4 == ed2);
   CHECK(id4 == id2);
+
+  SECTION("ParentageID unchanging") {
+    {
+      const std::string idString = "d41d8cd98f00b204e9800998ecf8427e";
+      std::string toString;
+      id1.toString(toString);
+      CHECK(toString == idString);
+    }
+
+    {
+      const std::string idString = "2e5751b7cfd7f053cd29e946fb2649a4";
+      std::string toString;
+      id2.toString(toString);
+      CHECK(toString == idString);
+    }
+    {
+      const std::string idString = "20e13ca818af45e50e369e50db3914b8";
+      std::string toString;
+      id3.toString(toString);
+      CHECK(toString == idString);
+    }
+    {
+      edm::Parentage ed_mult;
+      ed_mult.setParents(std::vector<edm::BranchID>({edm::BranchID(1), edm::BranchID(2), edm::BranchID(3)}));
+      auto id_mult = ed_mult.id();
+      const std::string idString = "6a5cf1697e50ec8e8dbe7a28ccad348b";
+      std::string toString;
+      id_mult.toString(toString);
+      CHECK(toString == idString);
+    }
+  }
 }

--- a/FWCore/Utilities/interface/Digest.h
+++ b/FWCore/Utilities/interface/Digest.h
@@ -5,6 +5,7 @@
 
 #include <iosfwd>
 #include <string>
+#include <string_view>
 #include <array>
 
 namespace cms {
@@ -48,9 +49,12 @@ namespace cms {
   public:
     Digest();
     explicit Digest(std::string const& s);
+    explicit Digest(std::string_view);
+    explicit Digest(const char*);
 
     void append(std::string const& s);
     void append(const char* data, size_t size);
+    void append(std::string_view v);
 
     MD5Result digest();
 

--- a/FWCore/Utilities/interface/Digest.h
+++ b/FWCore/Utilities/interface/Digest.h
@@ -28,7 +28,7 @@ namespace cms {
     std::string compactForm() const;
 
     // Set our data from the given hexdigest string.
-    void fromHexifiedString(std::string const& s);
+    void fromHexifiedString(std::string_view);
 
     bool isValid() const;
   };

--- a/FWCore/Utilities/src/Digest.cc
+++ b/FWCore/Utilities/src/Digest.cc
@@ -158,9 +158,24 @@ namespace cms {
     this->append(s);
   }
 
+  Digest::Digest(std::string_view v) : state_() {
+    md5_init(&state_);
+    this->append(v);
+  }
+
+  Digest::Digest(const char* s) : state_() {
+    md5_init(&state_);
+    this->append(s, strlen(s));
+  }
+
   void Digest::append(std::string const& s) {
     const md5_byte_t* data = reinterpret_cast<const md5_byte_t*>(s.data());
     md5_append(&state_, const_cast<md5_byte_t*>(data), s.size());
+  }
+
+  void Digest::append(std::string_view v) {
+    const md5_byte_t* data = reinterpret_cast<const md5_byte_t*>(v.data());
+    md5_append(&state_, const_cast<md5_byte_t*>(data), v.size());
   }
 
   void Digest::append(const char* s, size_t size) {

--- a/FWCore/Utilities/src/Digest.cc
+++ b/FWCore/Utilities/src/Digest.cc
@@ -114,13 +114,13 @@ namespace cms {
     return std::string(p, p + bytes.size());
   }
 
-  void MD5Result::fromHexifiedString(std::string const& hexy) {
+  void MD5Result::fromHexifiedString(std::string_view hexy) {
     switch (hexy.size()) {
       case 0: {
         set_to_default(*this);
       } break;
       case 32: {
-        std::string::const_iterator it = hexy.begin();
+        auto it = hexy.cbegin();
         for (size_t i = 0; i != 16; ++i) {
           // first nybble
           bytes[i] = (unhexify(*it++) << 4);

--- a/FWCore/Utilities/test/test_catch2_Digest.cc
+++ b/FWCore/Utilities/test/test_catch2_Digest.cc
@@ -26,7 +26,7 @@ namespace {
 TEST_CASE("Test cms::Digest", "[Digest]") {
   SECTION("Identical") {
     Digest dig1;
-    dig1.append("hello");
+    dig1.append(std::string_view("hello"));
     Digest dig2("hello");
 
     MD5Result r1 = dig1.digest();
@@ -80,5 +80,19 @@ TEST_CASE("Test cms::Digest", "[Digest]") {
         REQUIRE(lookup.compactForm() == fromHex.compactForm());
       }
     }
+  }
+  SECTION("append equal") {
+    std::string full("aldjfakl\tsdjf34234 \najdf");
+    Digest full_digest{full};
+    MD5Result full_r = full_digest.digest();
+    REQUIRE(full_r.isValid());
+
+    Digest append_digest;
+    append_digest.append(std::string_view(full.data(), 10));
+    append_digest.append(std::string_view(full.data() + 10, 10));
+    append_digest.append(std::string_view(full.data() + 20, 4));
+    MD5Result append_r = append_digest.digest();
+    REQUIRE(append_r.isValid());
+    REQUIRE(full_r == append_r);
   }
 }

